File: js/path_sum_2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
 var pathSum = function(root, targetSum) {
    return solve(root, targetSum, [], []);
};

function solve(root, sum, path, result) {
    if (!root) {
        return result;
    }

    path.push(root.val);

    if (!root.left && !root.right && root.val === sum) {
        result.push(path.slice());
    }

    solve(root.left, sum - root.val, path, result); 
    solve(root.right, sum - root.val, path, result);

    path.pop();

    return result;
}

File: java/path_sum_2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List <List <Integer>> ans =  new ArrayList<>();
        List <Integer> path = new ArrayList<>();
        
        solve(root, targetSum, ans, path);
        
        return ans;
    }
    
    public void solve(TreeNode root, int target, List <List <Integer>> ans, List <Integer> path) {
         if (root == null) {
            return;
        }
        
        path.add(root.val);
        if (target == root.val && root.left == null && root.right == null) {
            ans.add(new ArrayList<>(path));
        }
        
    
        solve(root.left, target-root.val, ans, path);
        solve(root.right, target-root.val, ans, path);
        path.remove(path.size()-1);
    }
}